python/ray/serve/api.py

```diff
@@ -1622,9 +1647,41 @@ def from_yaml(cls, str_or_file: Union[str, TextIO]) -> "Application":
                 Serve YAML config files.
 
         Returns:
-            Application: a new application object containing the deployments.
+            Application: a new Application object containing the deployments.
         """
-        raise NotImplementedError()
+
+        deployments_json = yaml.safe_load(str_or_file)
+        schema = ServeApplicationSchema.parse_obj(deployments_json)
+        return cls(schema_to_serve_application(schema))
+
+    def _add_deployment(self, deployment: Deployment):
+        """Adds the deployment to this Serve application.
+
+        Validates that a deployment with the same name doesn't already exist.
+        To overwrite an existing deployment, use index notation. For example:
+
+        app[deployment_name] = deployment
+
+        Args:
+            deployment (Deployment): deployment to add to this Application.
+
+        Raises:
+            ValueError: If a deployment with deployment.name already exists in
+                this Application.
+        """
+
+        if not isinstance(deployment, Deployment):
+            raise TypeError(f"Got {type(deployment)}. Expected deployment.")
+        elif deployment.name in self._deployments:
+            raise ValueError(
+                f'Deployment with name "{deployment.name}" already '
+                "exists in this application. To overwrite this "
+                "deployment, use attribute or index notation. "
+                "For example:\n\napp.deployment_name = "
+                "new_deployment"
+            )
+
```

Comment:
is this necessary? we should only support updating the config of existing deployments now, right?

Reply:
I think it's still useful, but the messages are outdated. The ValueError should prevent initializing the `app` using deployments with matching names.

Reply:
Actually, since it's never re-used, this logic can be pulled into the `__init__`. I'll make the change.